presto-main/src/main/java/com/facebook/presto/dispatcher/DispatchManager.java

```diff
@@ -193,6 +196,9 @@ private <C> void createQueryInternal(QueryId queryId, String slug, int retryCoun
             // decode session
             session = sessionSupplier.createSession(queryId, sessionContext, warningCollectorFactory);
 
+            // update identity if needed
+            session = updateIdentity(session);
```

Comment:
Can you leverage the sessionContext to populate the updatedIdentity? In that way you don't have to have extra methods in the dispatch manager.